### PR TITLE
8266179: [macos] jpackage should specify architecture for produced pkg files

### PR DESCRIPTION
**Pull request by sashamatveev**
jpackage should specify architecture for produced PKG files via hostArchitectures="x86_x64 or arm64". aarch64 installer will be installable on x64 without specifying hostArchitectures which is not correct and if install on arm Mac it will request Rosetta 2. With proposed fix by setting hostArchitectures="x86_x64" if installer contains x64 binaries, it will be installable on x64 Mac and will require Rosetta 2 on arm Mac. hostArchitectures will be set to arm64 if installer contain aarch64 binaries and will gave error when run on x64 Mac and will be installable on arm Mac without triggering installation of Rosetta 2.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266179](https://bugs.openjdk.java.net/browse/JDK-8266179): [macos] jpackage should specify architecture for produced pkg files

### Reviewers
 * [Andy Herrick](https://openjdk.java.net/census#herrick) (@andyherrick - **Reviewer**) ⚠️ Review applies to 1937e9e19856e63b5d4a3c83b3ac534018aa00d9
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - Author)
 * [Alexey Semenyuk](https://openjdk.java.net/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3807/head:pull/3807` \
`$ git checkout pull/3807`

Update a local copy of the PR: \
`$ git checkout pull/3807` \
`$ git pull https://git.openjdk.java.net/jdk pull/3807/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3807`

View PR using the GUI difftool: \
`$ git pr show -t 3807`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3807.diff">https://git.openjdk.java.net/jdk/pull/3807.diff</a>

</details>

### CODE REVIEW

**Review by kevinrushforth — APPROVED** [review]
The fix looks good. Would it be feasible to include an automated test for this?

**Review by alexeysemenyukoracle — APPROVED** [review]

**Comment by sashamatveev** [comment]
Added HostArchPkgTest test to verify hostArchitectures attribute.

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        if ("aarch64".equals(System.getProperty("os.arch"))) {
+            TKit.assertEquals(v, "arm64",
+                    "Check value of \"hostArchitectures\" attribute");
+        } else {
+            TKit.assertEquals(v, "x86_64",
+                    "Check value of \"hostArchitectures\" attribute");
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:84]: Please don't use direct TKit.run() call. Use jdk.jpackage.test.Annotations.Test annotation for test method. You can use SimplePackageTest jtreg test as an example

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * @requires (os.family == "mac")
+ * @key jpackagePlatformPackage
+ * @run main/othervm -Xmx512m HostArchPkgTest
+ */
+public class HostArchPkgTest {
+
+    private static void verifyHostArch(JPackageCommand cmd) throws Exception {
+        Path distributionFile = cmd.unpackedPackageDirectory()
+                .toAbsolutePath()
+                .getParent()
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:57]: Why .getParent() is needed here?

**Review by alexeysemenyukoracle — CHANGES_REQUESTED** [review]

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java (hunk 1 above)
Comment (sashamatveev) [reply test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:84]: I will fix it. Do we have a bug or I can file one to fix other tests? We have several tests such as SigningPackageTest which uses TKit.run() call and I just copy paste it.

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * @modules jdk.jpackage/jdk.jpackage.internal
+ * @compile HostArchPkgTest.java
+ * @requires (os.family == "mac")
+ * @key jpackagePlatformPackage
+ * @run main/othervm -Xmx512m HostArchPkgTest
+ */
+public class HostArchPkgTest {
+
+    private static void verifyHostArch(JPackageCommand cmd) throws Exception {
+        Path distributionFile = cmd.unpackedPackageDirectory()
```
[the remark this answers is not in the document]
Comment (sashamatveev) [reply test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:55]: Fixed.

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java (hunk 2 above)
Comment (sashamatveev) [reply test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:57]: Unpacking pkg is two stage process. First we unpack it to unpacked-pkg/data which contains Distribution file, pkg background images and tar archive with app itself. Then we unpack app tar archive to unpacked-pkg/unpacked. Test needs to check unpacked-pkg/data/Distribution and cmd.pathToUnpackedPackageFile("/") returns path to unpacked-pkg/unpacked. Thus getParent() is used to move up.

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java (hunk 1 above)
Comment (alexeysemenyukoracle) [reply test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:84]: No, we don't have a CR to track replacement of direct TKit.run() calls. Please feel free to file one.

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java (hunk 2 above)
Comment (alexeysemenyukoracle) [reply test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:57]: Understood. Thank you for explanation!

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Test
+    public static void test() {
+        new PackageTest()
+                .excludeTypes(PackageType.MAC_DMG) // Not valid case
+                .configureHelloApp()
+                .forTypes(PackageType.MAC_PKG)
+                .addInstallVerifier(HostArchPkgTest::verifyHostArch)
+                .run();
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:93]: The test is applicable only to the scenario when .pkg installer is unpacked and not when it is installed. So `PackageTest.run()` is not quite a good fit for this execution scenario as it depends on the value of `jpackage.test.action` system property (its default value is indeed to create and unpack installer, but can be overriden). The better option would be to use `PackageTest.run()` with explicit list of actions the test should perform. Suggested fix:
```
new PackageTest()
                .forTypes(PackageType.MAC_PKG)
                .configureHelloApp()
                .addInstallVerifier(HostArchPkgTest::verifyHostArch)
                .run(PackageTest.Action.CREATE_AND_UNPACK);
```

**Review by alexeysemenyukoracle — CHANGES_REQUESTED** [review]

File: test/jdk/tools/jpackage/macosx/HostArchPkgTest.java (hunk 4 above)
Comment (alexeysemenyukoracle) [reply test/jdk/tools/jpackage/macosx/HostArchPkgTest.java:93]: Sorry for the inconvenience, I didn't include this comment in my initial review.

**Review by alexeysemenyukoracle — APPROVED** [review]

**Comment by kevinrushforth** [comment]
I presume you've done a CI test build machines of both architectures?

**Comment by sashamatveev** [comment]
Yes, it was tested on both x64 and arm.

**Review by kevinrushforth — APPROVED** [review]

**Pull request closed without merging**